File: CFD_Principles_and_Applications_Blazek/BookCode/Fortran/structured_1d/srcmg/bcond.f
      SUBROUTINE Bcond( imax,ib2,a,cv,p )

C *****************************************************************************
C
C  Sets the boundary conditions at the inlet and outlet (in dummy points).
C
C *****************************************************************************
C
C  Subroutines called: none
C
C  Functions called: none
C
C *****************************************************************************
C
C  Created    : Jul. 25, 1997; (c) Jiri Blazek
C  Last update: Dec. 22, 1999
C
C  This program is free software; you can redistribute it and/or
C  modify it under the terms of the GNU General Public License
C  as published by the Free Software Foundation; either version 2
C  of the License, or (at your option) any later version.
C
C  This program is distributed in the hope that it will be useful,
C  but WITHOUT ANY WARRANTY; without even the implied warranty of
C  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
C  GNU General Public License for more details.
C
C  You should have received a copy of the GNU General Public License
C  along with this program; if not, write to the Free Software
C  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
C
C *****************************************************************************

      IMPLICIT NONE
      INCLUDE 'param.inc'

C ... parameter list
      INTEGER imax, ib2
      REAL*8 a(imax), cv(imax,3), p(imax)

C ... local variables
      REAL*8 u, cs2, c02, rinv, dis, cb, cc02,
     &       tb, pb, rhob, rho, cs, ub

C *****************************************************************************
C --- inlet (subsonic):
C
C     - speed of sound from Riemann invariant
C     - temperature from isentropic relation
C     - pressure from isentropic relation
C     - density from gas equation
C     - velocity from energy equation

      u    = cv(2,2)/cv(2,1)
      cs2  = gamma*p(2)*a(2)/cv(2,1)
      c02  = cs2 + 0.5D0*gam1*u*u
      rinv = u - 2.D0*SQRT(cs2)/gam1
      dis  = gap1*c02/(gam1*rinv*rinv) - 0.5D0*gam1
      IF (dis .LT. 0.D0) THEN
        WRITE(*,1000) dis
        dis = 1.D-20
      ENDIF
      cb   = -rinv*(gam1/gap1)*(1.D0+SQRT(dis))
      cc02 = MIN(cb*cb/c02, 1.D0)
      tb   = t01*cc02
      pb   = p01*(tb/t01)**(gamma/gam1)
      rhob = pb/(rgas*tb)
      ub   = SQRT(2.D0*cpgas*(t01-tb))

      cv(1,1) = rhob*a(2)
      cv(1,2) = rhob*a(2)*ub
      cv(1,3) = (pb/gam1+0.5D0*rhob*ub*ub)*a(2)
      p(1)    = pb

C --- outlet - subsonic:
C
C     - pressure = given back pressure
C     - density from characteristic b.c.
C     - velocity from characteristic b.c.
C
C --- outlet - supersonic:
C
C     - pressure extrapolated from the interior
C     - density extrapolated from the interior
C     - velocity extrapolated from the interior

      rho = cv(ib2,1)/a(ib2)
      u   = cv(ib2,2)/cv(ib2,1)
      cs  = SQRT(gamma*p(ib2)/rho)

      IF (u .GE. cs) THEN
C ----- supersonic flow
        pb   = p(ib2)
        rhob = rho
        ub   = u
      ELSE
C ----- subsonic flow
        pb   = p2
        rhob = rho + (p2-p(ib2))/(cs*cs)
        ub   = u   - (p2-p(ib2))/(cs*rho)
      ENDIF

      cv(imax,1) = rhob*a(ib2)
      cv(imax,2) = rhob*ub*a(ib2)
      cv(imax,3) = (pb/gam1+0.5D0*rhob*ub*ub)*a(ib2)
      p (imax)   = pb

1000  FORMAT(' Warning (bcond.f) - discriminant<0 at inflow boundary',
     &       ' (',1PE13.5,')')
      RETURN
      END

File: Aircraft_Aerodynamic_Design_with_Computational_Software_Rizzi/ExercisesLabs/Ch4/Ch4lib/bcond.m

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%                                                                   %
%                         DEMOFLOW                                  %
%                                                                   %
%       CALCULATION OF QUASI-1D INVISCID FLOW IN A TUBE/NOZZLE      %
%       ======================================================      %
%             Alexander von Essen, Created Apr. 11 2005             %
%                   Last modified: May 10 2005
%                   MG additions Levmax, W, p, Nodes, a {1:Levmax}  %
%                      aug 2017                                     %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%                                                                   %
% Name: bcond.m                                                     %
% Purpose: Sets boundary conditions in dummy points.                %
% Called by: main.m solver.m                                        %
% Calls:                                                            %
%                                                                   %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
function handles = bcond(handles)
%W,p = f(p01,t01,p2,gamma,R; Nodes,a,p,W)

p01   = handles.Data.p01;
t01   = handles.Data.t01;
p2    = handles.Data.p2;
gamma = handles.Data.gamma;
R     = handles.Data.R;
cpgas = R*gamma/(gamma-1);

ib2 = handles.Data.Nodes-1;
a   = handles.Data.a;
p   = handles.Data.p;
W   = handles.Data.W;

% Inlet.
u    = W(2,2)/W(2,1);
cs2  = gamma*p(2)*a(2)/W(2,1);
c02  = cs2 + 0.5*(gamma-1)*u^2;
rinv = u - 2*sqrt(cs2)/(gamma-1);
dis  = (gamma+1)*c02/((gamma-1)*rinv^2) - 0.5*(gamma-1);
if dis < 0
    dis = 1E-20;
end
cb   = -rinv*((gamma-1)/(gamma+1))*(1+sqrt(dis));
cc02 = min(cb^2/c02,1);
tb   = t01*cc02;
pb   = p01*(tb/t01)^(gamma/(gamma-1));
rhob = pb/(R*tb);
ub   = sqrt(2*cpgas*(t01-tb));
% JO 2021
rho01 = p01/t01/R;
Htot  = gamma/(gamma-1)*p01/rho01;
c     = (-rinv*(gamma-1) + sqrt((gamma^2-1)*(Htot*(gamma-1) + rinv^2)))/(gamma+1);
S     = p01/rho01^gamma;
%rhob  = (c^2/gamma/S)^1/(gamma-1);
%pb    = S*rhob^gamma;
%ub    = rinv+2*c/(gamma-1);
W(1,1) = rhob*a(2);
W(1,2) = rhob*a(2)*ub;
W(1,3) = (pb/(gamma-1)+0.5*rhob*ub^2)*a(2);
p(1)   = pb;


% Outlet.
rho = W(ib2,1)/a(ib2);
u   = W(ib2,2)/W(ib2,1);
cs  = sqrt(gamma*p(ib2)/rho);
% JO 2104 - check that end pressure > p2
if u >= cs & p(ib2) > p2
    % Supersonic.
    pb   = p(ib2);
    rhob = rho;
    ub   = u;
else
    % Subsonic.
    pb   = p2;
    rhob = rho + (p2-p(ib2))/cs^2;
    ub   = u - (p2-p(ib2))/(cs*rho);
end
W(end,1) = rhob*a(ib2);
W(end,2) = rhob*ub*a(ib2);
W(end,3) = (pb/(gamma-1)+0.5*rhob*ub^2)*a(ib2);
p(end)   = pb;

handles.Data.p = p;
handles.Data.W = W;
